File: packages/ios/Sources/Components/Button/Button.swift
import SwiftUI

public enum KozmosButtonVariant {
    case `default`
    case destructive
    case outline
    case secondary
    case ghost
    case link
    case glass
}

public enum KozmosButtonSize {
    case `default`
    case sm
    case lg
    case icon
}

public enum KozmosButtonEmotion {
    case themed
    case neutral
    case success
    case danger
    case informative
    case alert
}

/// Which token tier a variant's shape belongs to. Primary is filled,
/// Secondary is bordered or text; `glass` is an effect and takes no emotion.
enum KozmosButtonTier {
    case primary
    case secondary
}

public struct KozmosButton: View {
    let label: String
    let variant: KozmosButtonVariant
    let emotion: KozmosButtonEmotion?
    let size: KozmosButtonSize
    let isDisabled: Bool
    let isLoading: Bool
    let action: () -> Void
    
    public init(
        _ label: String,
        variant: KozmosButtonVariant = .default,
        emotion: KozmosButtonEmotion? = nil,
        size: KozmosButtonSize = .default,
        isDisabled: Bool = false,
        isLoading: Bool = false,
        action: @escaping () -> Void
    ) {
        self.label = label
        self.variant = variant
        self.emotion = emotion
        self.size = size
        self.isDisabled = isDisabled
        self.isLoading = isLoading
        self.action = action
    }
    
    public var body: some View {
        Button(action: action) {
            HStack(spacing: KozmosDimensions.primitivesLayoutSpacing100) {
                if isLoading {
                    ProgressView()
                        .controlSize(.small)
                        .tint(foregroundColor)
                }
                Text(label)
                    .font(KozmosTypography.subheadline)
                    .fontWeight(.medium)
            }
            .padding(padding)
            .foregroundColor(foregroundColor)
            .frame(
                minWidth: size == .icon ? 44 : nil,
                minHeight: 44
            )
            .kozmosButtonSurface(
                variant == .glass ? .glass : nil,
                fill: backgroundColor,
                stroke: borderColor,
                strokeWidth: variant == .outline ? 1 : 0,
                // The corner the button always had; a continuous one would move every baseline.
                shape: RoundedRectangle(cornerRadius: KozmosDimensions.semanticsRadiusControl)
            )
        }
        .disabled(isDisabled || isLoading)
        .opacity(isDisabled ? 0.5 : 1)
    }
    

    

    /// The tier this variant reads, or nil when the variant is an effect.
    private var tier: KozmosButtonTier? {
        switch variant {
        case .default, .secondary, .destructive: return .primary
        case .outline, .ghost, .link: return .secondary
        case .glass: return nil
        }
    }

    private static func emotionBackground(_ tier: KozmosButtonTier, _ emotion: KozmosButtonEmotion) -> Color {
        switch (tier, emotion) {
        case (.primary, .themed): return KozmosColors.componentsPrimaryButtonsThemedButtonBackgroundIdle
        case (.primary, .neutral): return KozmosColors.componentsPrimaryButtonsNeutralButtonBackgroundIdle
        case (.primary, .success): return KozmosColors.componentsPrimaryButtonsSuccessButtonBackgroundIdle
        case (.primary, .danger): return KozmosColors.componentsPrimaryButtonsDangerButtonBackgroundIdle
        case (.primary, .informative): return KozmosColors.componentsPrimaryButtonsInformativeButtonBackgroundIdle
        case (.primary, .alert): return KozmosColors.componentsPrimaryButtonsAlertButtonBackgroundIdle
        case (.secondary, .themed): return KozmosColors.componentsSecondaryButtonsThemedButtonBackgroundIdle
        case (.secondary, .neutral): return KozmosColors.componentsSecondaryButtonsNeutralButtonBackgroundIdle
        case (.secondary, .success): return KozmosColors.componentsSecondaryButtonsSuccessButtonBackgroundIdle
        case (.secondary, .danger): return KozmosColors.componentsSecondaryButtonsDangerButtonBackgroundIdle
        case (.secondary, .informative): return KozmosColors.componentsSecondaryButtonsInformativeButtonBackgroundIdle
        case (.secondary, .alert): return KozmosColors.componentsSecondaryButtonsAlertButtonBackgroundIdle
        }
    }

    private static func emotionForeground(_ tier: KozmosButtonTier, _ emotion: KozmosButtonEmotion) -> Color {
        switch (tier, emotion) {
        case (.primary, .themed): return KozmosColors.componentsPrimaryButtonsThemedButtonForegroundContentIdle
        case (.primary, .neutral): return KozmosColors.componentsPrimaryButtonsNeutralButtonForegroundContentIdle
        case (.primary, .success): return KozmosColors.componentsPrimaryButtonsSuccessButtonForegroundContentIdle
        case (.primary, .danger): return KozmosColors.componentsPrimaryButtonsDangerButtonForegroundContentIdle
        case (.primary, .informative): return KozmosColors.componentsPrimaryButtonsInformativeButtonForegroundContentIdle
        case (.primary, .alert): return KozmosColors.componentsPrimaryButtonsAlertButtonForegroundContentIdle
        case (.secondary, .themed): return KozmosColors.componentsSecondaryButtonsThemedButtonForegroundContentIdle
        case (.secondary, .neutral): return KozmosColors.componentsSecondaryButtonsNeutralButtonForegroundContentIdle
        case (.secondary, .success): return KozmosColors.componentsSecondaryButtonsSuccessButtonForegroundContentIdle
        case (.secondary, .danger): return KozmosColors.componentsSecondaryButtonsDangerButtonForegroundContentIdle
        case (.secondary, .informative): return KozmosColors.componentsSecondaryButtonsInformativeButtonForegroundContentIdle
        case (.secondary, .alert): return KozmosColors.componentsSecondaryButtonsAlertButtonForegroundContentIdle
        }
    }

    private var padding: EdgeInsets {
        switch size {
        case .default: return EdgeInsets(top: 0, leading: KozmosDimensions.primitivesLayoutSpacing200, bottom: 0, trailing: KozmosDimensions.primitivesLayoutSpacing200)
        case .sm: return EdgeInsets(top: 0, leading: KozmosDimensions.primitivesLayoutSpacing150, bottom: 0, trailing: KozmosDimensions.primitivesLayoutSpacing150)
        case .lg: return EdgeInsets(top: 0, leading: KozmosDimensions.primitivesLayoutSpacing400, bottom: 0, trailing: KozmosDimensions.primitivesLayoutSpacing400)
        case .icon: return EdgeInsets(top: 0, leading: 0, bottom: 0, trailing: 0)
        }
    }
    
    private var backgroundColor: Color {
        if let emotion, let tier {
            // A bordered or text button keeps its transparent ground; only a
            // filled one takes the emotion's background.
            switch variant {
            case .outline, .ghost, .link: return Color.clear
            default: return Self.emotionBackground(tier, emotion)
            }
        }
        switch variant {
        case .default: return KozmosColors.componentsPrimaryButtonsThemedButtonBackgroundIdle // Used specific component token
        case .destructive: return KozmosColors.componentsPrimaryButtonsDangerButtonBackgroundIdle
        case .outline, .ghost: return Color.clear
        case .secondary: return KozmosColors.componentsPrimaryButtonsNeutralButtonBackgroundIdle
        case .link: return Color.clear
        // The glass variant is the glass surface; the fill is the surface's.
        case .glass: return Color.clear
        }
    }
    
    private var foregroundColor: Color {
        if let emotion, let tier {
            return Self.emotionForeground(tier, emotion)
        }
        switch variant {
        case .default: return KozmosColors.componentsPrimaryButtonsThemedButtonForegroundContentIdle // Semantic Token
        case .destructive: return KozmosColors.componentsPrimaryButtonsDangerButtonForegroundContentIdle
        case .outline, .ghost, .link: return KozmosColors.primitivesColorsTheme500
        case .secondary: return KozmosColors.primitivesColorsForeground100
        case .glass: return KozmosColors.primitivesColorsForeground100
        }
    }
    
    private var borderColor: Color {
        if let emotion, let tier, variant == .outline {
            return Self.emotionForeground(tier, emotion)
        }
        switch variant {
        case .outline: return KozmosColors.primitivesColorsForeground300
        default: return Color.clear
        }
    }
}

extension View {
    /// A button's surface: the glass surface role for the glass variant —
    /// composed from `Semantics.Effect.glass`, as every glass surface is —
    /// or the variant's own fill and edge.
    @ViewBuilder
    func kozmosButtonSurface<S: InsettableShape>(
        _ style: KozmosSurfaceStyle?,
        fill: Color,
        stroke: Color,
        strokeWidth: CGFloat,
        shape: S
    ) -> some View {
        if let style {
            kozmosSurface(shape, style: style)
        } else {
            background(fill)
                .clipShape(shape)
                .overlay(shape.stroke(stroke, lineWidth: strokeWidth))
        }
    }
}

File: packages/android/src/main/java/com/kozmos/components/Button/Button.kt
package com.kozmos.components.button

import androidx.compose.foundation.BorderStroke
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Button
import androidx.compose.material3.ButtonDefaults
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.LocalContentColor
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.OutlinedButton
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.runtime.Composable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.RowScope
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.unit.dp
import androidx.compose.ui.graphics.Color
import com.kozmos.tokens.KozmosDimensions
import com.kozmos.components.surface.KozmosSurfaceDefaults
import com.kozmos.components.surface.KozmosSurfaceStyle
import com.kozmos.tokens.KozmosThemeTokens

enum class KozmosButtonVariant {
    Default,
    Destructive,
    Outline,
    Secondary,
    Ghost,
    Link,
    Glass
}

enum class KozmosButtonEmotion {
    Themed,
    Neutral,
    Success,
    Danger,
    Informative,
    Alert
}

enum class KozmosButtonSize {
    Default,
    Sm,
    Lg,
    Icon
}

// The emotion decides the colour where the variant has a tier: Primary is
// filled, Secondary is bordered or text. Glass is an effect and keeps its
// own ground.
@Composable
private fun emotionPrimaryBackground(e: KozmosButtonEmotion): Color = when (e) {
    KozmosButtonEmotion.Themed -> KozmosThemeTokens.componentsPrimaryButtonsThemedButtonBackgroundIdle
    KozmosButtonEmotion.Neutral -> KozmosThemeTokens.componentsPrimaryButtonsNeutralButtonBackgroundIdle
    KozmosButtonEmotion.Success -> KozmosThemeTokens.componentsPrimaryButtonsSuccessButtonBackgroundIdle
    KozmosButtonEmotion.Danger -> KozmosThemeTokens.componentsPrimaryButtonsDangerButtonBackgroundIdle
    KozmosButtonEmotion.Informative -> KozmosThemeTokens.componentsPrimaryButtonsInformativeButtonBackgroundIdle
    KozmosButtonEmotion.Alert -> KozmosThemeTokens.componentsPrimaryButtonsAlertButtonBackgroundIdle
}

@Composable
private fun emotionPrimaryForeground(e: KozmosButtonEmotion): Color = when (e) {
    KozmosButtonEmotion.Themed -> KozmosThemeTokens.componentsPrimaryButtonsThemedButtonForegroundContentIdle
    KozmosButtonEmotion.Neutral -> KozmosThemeTokens.componentsPrimaryButtonsNeutralButtonForegroundContentIdle
    KozmosButtonEmotion.Success -> KozmosThemeTokens.componentsPrimaryButtonsSuccessButtonForegroundContentIdle
    KozmosButtonEmotion.Danger -> KozmosThemeTokens.componentsPrimaryButtonsDangerButtonForegroundContentIdle
    KozmosButtonEmotion.Informative -> KozmosThemeTokens.componentsPrimaryButtonsInformativeButtonForegroundContentIdle
    KozmosButtonEmotion.Alert -> KozmosThemeTokens.componentsPrimaryButtonsAlertButtonForegroundContentIdle
}

@Composable
private fun emotionSecondaryForeground(e: KozmosButtonEmotion): Color = when (e) {
    KozmosButtonEmotion.Themed -> KozmosThemeTokens.componentsSecondaryButtonsThemedButtonForegroundContentIdle
    KozmosButtonEmotion.Neutral -> KozmosThemeTokens.componentsSecondaryButtonsNeutralButtonForegroundContentIdle
    KozmosButtonEmotion.Success -> KozmosThemeTokens.componentsSecondaryButtonsSuccessButtonForegroundContentIdle
    KozmosButtonEmotion.Danger -> KozmosThemeTokens.componentsSecondaryButtonsDangerButtonForegroundContentIdle
    KozmosButtonEmotion.Informative -> KozmosThemeTokens.componentsSecondaryButtonsInformativeButtonForegroundContentIdle
    KozmosButtonEmotion.Alert -> KozmosThemeTokens.componentsSecondaryButtonsAlertButtonForegroundContentIdle
}

@Composable
fun KozmosButton(
    onClick: () -> Unit,
    modifier: Modifier = Modifier,
    variant: KozmosButtonVariant = KozmosButtonVariant.Default,
    emotion: KozmosButtonEmotion? = null,
    size: KozmosButtonSize = KozmosButtonSize.Default,
    enabled: Boolean = true,
    isLoading: Boolean = false,
    content: @Composable RowScope.() -> Unit
) {
    val height = when (size) {
        KozmosButtonSize.Default -> 44.dp
        KozmosButtonSize.Sm -> 44.dp
        KozmosButtonSize.Lg -> 44.dp
        KozmosButtonSize.Icon -> 44.dp
    }

    val rootModifier = modifier
        .height(height)
        .then(if (size == KozmosButtonSize.Icon) Modifier.width(44.dp) else Modifier)
    
    val contentPadding = when (size) {
        KozmosButtonSize.Default -> PaddingValues(horizontal = KozmosDimensions.primitivesLayoutSpacing200, vertical = 0.dp)
        KozmosButtonSize.Sm -> PaddingValues(horizontal = KozmosDimensions.primitivesLayoutSpacing150, vertical = 0.dp)
        KozmosButtonSize.Lg -> PaddingValues(horizontal = KozmosDimensions.primitivesLayoutSpacing400, vertical = 0.dp)
        KozmosButtonSize.Icon -> PaddingValues(0.dp)
    }

    // Dimension token graph synced from build.mjs
    val shape = RoundedCornerShape(KozmosDimensions.semanticsRadiusControl)

    when (variant) {
        KozmosButtonVariant.Outline -> {
            val outlineColor = if (emotion != null) emotionSecondaryForeground(emotion)
                else KozmosThemeTokens.componentsSecondaryButtonsThemedButtonForegroundContentIdle
            OutlinedButton(
                onClick = onClick,
                modifier = rootModifier,
                enabled = enabled && !isLoading,
                shape = shape,
                contentPadding = contentPadding,
                border = BorderStroke(1.dp, outlineColor),
                colors = ButtonDefaults.outlinedButtonColors(contentColor = outlineColor)
            ) {
                ButtonContent(isLoading, content)
            }
        }
        KozmosButtonVariant.Ghost, KozmosButtonVariant.Link -> {
            val textColor = if (emotion != null) emotionSecondaryForeground(emotion)
                else KozmosThemeTokens.componentsSecondaryButtonsThemedButtonForegroundContentIdle
            TextButton(
                onClick = onClick,
                modifier = rootModifier,
                enabled = enabled && !isLoading,
                shape = shape,
                contentPadding = contentPadding,
                colors = ButtonDefaults.textButtonColors(contentColor = textColor)
            ) {
                ButtonContent(isLoading, content)
            }
        }
        else -> {
            // Default, Destructive, Secondary, Glass
            val containerColor = if (emotion != null && variant != KozmosButtonVariant.Glass) emotionPrimaryBackground(emotion) else when(variant) {
                KozmosButtonVariant.Destructive -> KozmosThemeTokens.componentsPrimaryButtonsDangerButtonBackgroundIdle
                KozmosButtonVariant.Secondary -> KozmosThemeTokens.componentsPrimaryButtonsNeutralButtonBackgroundIdle
                // The glass variant is the glass surface, composed from the token.
                KozmosButtonVariant.Glass -> KozmosSurfaceDefaults.tint(KozmosSurfaceStyle.Glass)
                else -> KozmosThemeTokens.componentsPrimaryButtonsThemedButtonBackgroundIdle
            }
            
            val contentColor = if (emotion != null && variant != KozmosButtonVariant.Glass) emotionPrimaryForeground(emotion) else when(variant) {
                 KozmosButtonVariant.Secondary -> KozmosThemeTokens.componentsPrimaryButtonsNeutralButtonForegroundContentIdle
                 KozmosButtonVariant.Destructive -> KozmosThemeTokens.componentsPrimaryButtonsDangerButtonForegroundContentIdle
                 KozmosButtonVariant.Glass -> KozmosThemeTokens.primitivesColorsForeground100
                 else -> KozmosThemeTokens.componentsPrimaryButtonsThemedButtonForegroundContentIdle
            }
            
            Button(
                onClick = onClick,
                modifier = rootModifier,
                enabled = enabled && !isLoading,
                shape = shape,
                contentPadding = contentPadding,
                colors = ButtonDefaults.buttonColors(
                    containerColor = containerColor,
                    contentColor = contentColor
                ),
                border = if (variant == KozmosButtonVariant.Glass) KozmosSurfaceDefaults.border(KozmosSurfaceStyle.Glass) else null
            ) {
                ButtonContent(isLoading, content)
            }
        }
    }
}

@Composable
private fun RowScope.ButtonContent(
    isLoading: Boolean,
    content: @Composable RowScope.() -> Unit
) {
    // GAP-56: the loader and the caller's children keep 8 apart, as Figma's Button (itemSpacing
    // 8) and iOS's (HStack spacing 100) do. Both used to touch the label.
    Row(
        horizontalArrangement = Arrangement.spacedBy(KozmosDimensions.primitivesLayoutSpacing100),
        verticalAlignment = Alignment.CenterVertically,
    ) {
        if (isLoading) {
            CircularProgressIndicator(
                modifier = Modifier.size(14.dp),
                color = LocalContentColor.current,
                strokeWidth = 2.dp
            )
        }
        content()
    }
}
